server/src/main/java/org/elasticsearch/action/bulk/BulkRequestBuilder.java

```diff
@@ -199,11 +201,18 @@ public BulkRequestBuilder setRefreshPolicy(String refreshPolicy) {
 
     @Override
     public BulkRequest request() {
+        assert requestPreviouslyCalled == false : "Cannot call request() multiple times on the same BulkRequestBuilder object";
+        if (requestPreviouslyCalled) {
+            throw new IllegalStateException("Cannot call request() multiple times on the same BulkRequestBuilder object");
+        }
+        requestPreviouslyCalled = true;
         validate();
         BulkRequest request = new BulkRequest(globalIndex);
-        for (ActionRequestLazyBuilder<? extends DocWriteRequest<?>, ? extends DocWriteResponse> requestBuilder : requestBuilders) {
-            DocWriteRequest<?> childRequest = requestBuilder.request();
+        for (Iterator<ActionRequestLazyBuilder<? extends DocWriteRequest<?>, ? extends DocWriteResponse>> requestsIter = requestBuilders
+            .iterator(); requestsIter.hasNext();) {
+            DocWriteRequest<?> childRequest = requestsIter.next().request();
             request.add(childRequest);
+            requestsIter.remove(); // The inner request builder can now be garbage collected
```

Comment:
> Or maybe use a queue instead?

Yeah `ArrayDeque` is essentially an `ArrayList` with a moveable head so it'd work better here.

Reply:
I think you can avoid explicitly using the iterator and just clear the `ArrayList` after iterating over it.
```java
for (var requestBuilder : requestBuilders) {
    request.add(requestBuilder.request());
}
requestBuilders.clear();
```

Reply:
Waiting until the end and then clearing all the builders still has the 2x peak memory usage problem that we're trying to solve.

Reply:
Good point. I'll rework this. According to the ArrayDeque javadocs
> Most ArrayDeque operations run in amortized constant time. Exceptions include remove, removeFirstOccurrence,  removeLastOccurrence, contains, iterator.remove(), and the bulk operations, all of which run in linear time.

So it might not be the way to go either. Maybe null and clear (or new list -- seems like i've seen clear be slow for large lists) is the way to go.

Reply:
But you wouldn't need to use any of those O(N) operations with `ArrayDeque`, you'd be calling `pollFirst()` which is definitely O(1).

Reply:
Oops,  right. :)